sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataReadRunner.java

```diff
@@ -166,14 +177,32 @@ public void registerInputLocation() {
             apiServiceDescriptor,
             LogicalEndpoint.of(processBundleInstructionIdSupplier.get(), inputTarget),
             coder,
-            receiver);
+            enqueueElementReceiver);
   }
 
   public void blockTillReadFinishes() throws Exception {
     LOG.debug(
         "Waiting for process bundle instruction {} and target {} to close.",
         processBundleInstructionIdSupplier.get(),
         inputTarget);
-    readFuture.awaitCompletion();
+
+    boolean draining = false;
+    while (true) {
+      WindowedValue<OutputT> val = this.queue.poll(50, TimeUnit.MILLISECONDS);
+      if (val != null) {
+        // Call receiver.
+        elementReceiver.accept(val);
```

Comment:
Any exception that happened on the receiver stream would propagate up to the caller since it was happening on the receivers thread and that would cancel the stream. You'll want to make sure that the InboundDataClient is failed if processing fails. Please add a test for this case as well to the BeamFnDataReadRunner suite.

Also add a test for making sure that the receiver after the BeamFnDataClient is happening on the same tread as the start/finish functions.

Reply:
Added this type of test to the RemoteExecutionTest